### PR TITLE
[calyx] Change FSM size to variable dependent on number of states

### PR DESCRIPTION
**Pull request by cgyurgyik**
Reduces the FSM size from a constant 32 to a variable size determined by the number of states. This is done using a function `get_bit_width(N)`, determines the maximum number of bits to represent N states.

### CODE REVIEW

**Comment by sampsyo** [comment]
Cool! I couldn't resist writing an integer log function ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f8fdc16b8dfdb668320d81a39dc3b323))… I hope this is right! 😇

```rust
fn log2(n: u64) -> u8 {
    for i in 0u8..63 {
        if n & (1u64 << (63 - i)) != 0 {
            return 63 - i;
        }
    }
    panic!();
}
```

**Comment by cgyurgyik** [comment]
Thanks! That's what I was thinking about doing when refactoring to a single function. One nit, to be specific about whether its `floor_log2` or `ceil_log2`. If this ends up being a route we take I'll be sure to write tests as well.

**Comment by rachitnigam** [comment]
Do you know why tests are failing?

**Comment by cgyurgyik** [comment]
> Do you know why tests are failing?

For the formatting test, I think I just forgot to format some small change; this should be easily fixed.

For the build, not yet, I'll have to look into it. My first guess is that the tests are hardcoded to having FSMs with size 32, and my second guess is because this naive method doesn't work.

**Comment by rachitnigam** [comment]
The golden tests do expect to see the constants so that is probably one issue. From a quick glance, the simulation tests, which actually run the generated verilog, seem to work.

To fix the tests, you can test and run them locally using [runt](https://github.com/rachitnigam/runt). Ask questions about it in the channel.

**Comment by cgyurgyik** [comment]
Going to wait for #228 to be pushed through before making changes to tests.

**Comment by cgyurgyik** [comment]
I decided to take a slightly different approach rather than using log2. This acts similar to what ```log2``` would do, except for a few corner cases. For example, ```log2(1) is 0```. We don't want that. If the number of states in the FSM is 1, then we still want 1 bit. 

The function name is then updated to ```get_bit_width_from``` to match its intended use case, and acts as one might expect e.g., ```get_bit_width_from(0) == 0```.
Happy to take other name suggestions :)

**Comment by rachitnigam** [comment]
@cgyurgyik The new tests are merged in! Feel free to update this and request a review on it.

**Comment by cgyurgyik** [comment]
@rachitnigam @sgpthomas 
PTAL. I've updated the current tests to reflect the impact of this PR. With each `fsm_size` change, I've added some comments behind some of the "magic numbers" (e.g. need to add `1` for the `0`'th state).

File: calyx/src/passes/static_timing.rs [hunk 1]
```diff
@@ -90,7 +91,8 @@ impl Visitor for StaticTiming {
                 let while_group_node =
                     st.insert_group(&while_group, HashMap::new())?;
 
-                let fsm_size = 32;
+                // `0` state + (2 + btime) states, where (2 + btime) == fsm_loop_exit_state
```
Comment (rachitnigam) [inline calyx/src/passes/static_timing.rs:94]: Can you explain why you need the `2` here? What states does that correspond to?

**Comment by rachitnigam** [comment]
I just had one inline question.

File: calyx/src/passes/static_timing.rs (hunk 1 above)
Comment (cgyurgyik) [reply calyx/src/passes/static_timing.rs:94]: https://github.com/cucapra/futil/blob/e1eb074d0d2aa5f22b02ee5444323949e95dbda8/calyx/src/passes/static_timing.rs#L99

Comment (rachitnigam) [reply calyx/src/passes/static_timing.rs:94]: Oh wow, I didn't even know there was this black magic. @sgpthomas needs to stare at this to make sure its correct.

Comment (cgyurgyik) [reply calyx/src/passes/static_timing.rs:94]: Yeah one thing I could do is declare these variables before `fsm_size` as well, so that we don't need to change code in two places. I'll wait for Sam to respond before I make any changes.

Comment (sgpthomas) [reply calyx/src/passes/static_timing.rs:94]: For a body that takes n cycles, we use state 0 for the loop enter state, states 1 to n+1 for executing the body and n+2 for the loop exit state. It explains the state in the comment at the top of the while compilation function. Although I think there's an off by one typo in that. the correct comment should be:
```
                // FSM Encoding:
                //   0:     init state. we haven't started loop iterations
                //          and haven't checked the loop body
                //   1-n+1: body compute states. cond was true. compute the body
                //   n+2:   loop exit. we've finished running the body and the condition is false.
                // Transitions:
                //   0 -> 1:     when cond == true
                //   0 -> n+2:   when cond == false
                //   i -> i+1:   when i != 0 & i != n+1
                //   n+1-> 1:    when cond == true
                //   n+1 -> n+2: when cond == false
```

Comment (sgpthomas) [reply calyx/src/passes/static_timing.rs:94]: jk, this is the correct one:
```
                // FSM Encoding:
                //   0:   init state. we haven't started loop iterations
                //        and haven't checked the loop body
                //   1-n: body compute states. cond was true. compute the body
                //   n+1: check condition again
                //   n+2: loop exit. we've finished running the body and the condition is false.
                // Transitions:
                //   0   -> 1:   when cond == true
                //   0   -> n+2: when cond == false
                //   i   -> i+1: when i != 0 & i != n
                //   n+1 -> 1:   when cond == true
                //   n+1 -> n+2: when cond == false
```

Comment (cgyurgyik) [reply calyx/src/passes/static_timing.rs:94]: Huh ok never saw that comment. I'll need to double check my numbers against this, and I'll update the comment as well. I'll resolve this when I'm done.

**Review by sgpthomas — APPROVED** [review]

**Comment by rachitnigam** [comment]
Huh, looks like change somehow got clobbered.
